### ISSUE TITLE
Should env.reset() always copyt the model?

### ISSUE THREAD

**Issue by AntoinePrv** [ds4dm/ecole#103]
From @dchetelat 

> I have a weird issue with the new instance generators. I can't solve the same problem twice in a row?
```python
> instance = next(instances)
> default_env.reset(instance)
> _, _, _, _, default_info = default_env.step({})
> print(default_info)
{'nb_nodes': 1.0, 'time': 0.97}
> default_env.reset(instance)
> _, _, _, _, default_info = default_env.step({})
> print(default_info)
{'nb_nodes': 0.0, 'time': 0.0}
```

> if it's not too costly, I think it should be done by default by reset, the use case I describe above will be extremely common and it's rather counterintuitive that I have to do .copy_orig()

**Comment by AntoinePrv** [comment]
On thing is that if a copy is made in `reset`, then users who do not want a copy, or who may like another copy functions from SCIP, cannot have it anymore, whereas with no copy in `reset`, it is still possible to make one.
In this case _less is more_.

**Comment by gasse** [comment]
The way I see it, from a user perspective, instance generators return instances, not SCIP models. We use a SCIP model internally as a convenient container to store the instance specification (original problem in SCIP), and because it directly offers useful utilities such as `write_problem` to save the instance to an LP file. Didier's use case is natural I think: load an instance in an environment, then solve it. Then load the instance again, in a different environment, and solve it. Having to call `instance = instance.copy_orig()` for users to have this behavior is not very intuitive, as it requires users to actually be aware that an instance is not an immutable instance object, but a SCIP model that will be altered by any environment it is passed to.

@AntoinePrv what would be a scenario where a user does not want a copy ? I could see two reasons for not wanting a copy in `env.reset()`:

1. the memory or the computation cost of a copy
2. if users want to read/write model properties before or after `env.reset()` or `env.step()`

I am not sure what the cost of 1. would be in practice, maybe it would have an impact for very large instances ?

I don't think we really want users to interact with the SCIP model via the `instance` object, as it potentially messes up with the PO-MDP API, which is the intended use of Ecole.

What do you think guys ?

**Comment by dchetelat** [comment]
Yeah I'm really in agreement with @gasse here. For me, as a user, the instance is the abstract problem, the (A, b, c) in the linear formulation. If you take this point of view, which is I think the point of view most researchers would have, then this is somewhat incompatible with with how environments would modify instances: if an instance=(A, b, c), then why exactly would solving my instance, i.e. doing
```python
env1.reset(instance)
while not done:
    env1.step(action)
```
change my (A, b, c)=instance? Worse, this idempotence would differ based on whether I'm getting, say, the instance from a file vs. from a generator... With this (A, b, c) perspective, which is I think the perspective most practitioners would have, then it is a kind of failure that the instances are sometimes carrying more than just the problem formulation (like SCIP buffers and everything).

So I would disagree with the less is more motto. By not making sure that environment solving is idempotent with respect to instances, we are adding additional, counterintuitive concerns to users, for example that they can perfectly do
```python
env1.reset(instance)
env1.reset(instance)
```
when instance is coming from a file, but that this won't work if it comes from a generator, whereas it is supposed to contain the same info, the (A, b, c)! It might be simpler the developers, but it's more brainwork for our target users...

And I think for no real good reason either, because I don't really see what use case automatic copying in reset would prevent. Perhaps the only use case I can think of is, if you do something like
```python
env1.reset(instance)
while not done:
    env1.step(action)

# do some stuff on the model with pyscipopt
model = env.model.as_pyscipopt()
# ...

env2.reset(instance)
# ...
```
and the modifications brought by the pyscipopt modifications on the model of the first environment would not be carried over to the second environment. But there are easy fixes for this (calling `env2.reset(model)` say) and anyways, in my mind if you are doing some pyscipopt you are already doing something weird, so it's normal and expected you might have to do some less natural code.

So for me, the only disadvantage I can think of is speed/memory, the additional cost of making a copy. My suspicion is that it's rather negligible, and I think we can perfectly live with that.

---

I should say, also: I think in some sense, the real solution here would be to bridge better the gap between the user and the programming concepts. Namely, perhaps the better solution here would be that the generators output an intermediate `Instance` class that only contains (A, b, c), and environments would turn that into a SCIP model in reset. That way it would match better what the scientists think an instance is. But this would represent quite some rework of the instance generators, so, I think just adding a copying step in the reset is a simple fix that would work well at this point. For sure it would allow us to ship a clean branching tutorial in time for the Neurips workshop.

**Comment by AntoinePrv** [comment]
> The way I see it, from a user perspective, instance generators return instances, not SCIP models.

> For me, as a user, the instance is the abstract problem, the (A, b, c) in the linear formulation.

I understand, this makes sense. In fact, `env.reset` should also not accept `scip.Model` because many of these states would be undefined for the environment.

> the memory or the computation cost of a copy

> So for me, the only disadvantage I can think of is speed/memory, the additional cost of making a copy when solving once. My suspicion is that it's rather negligible,

I also think it is more than acceptable (in general I still try to avoid even acceptable costs, as when they accumulate they may  not stay acceptable).

> Namely, perhaps the better solution here would be that the generators output an intermediate Instance class that only contains (A, b, c), and environments would turn that into a SCIP model in reset. 

I agree, however I don't believe we should build that class now. We have bigger priorities, and this is something we will always be lagging behind when the type of instance we want to resolve would get more complicated (e.g. non linear).

On thing that we can perhaps more easily do is have an `Instance` class that also utilize a `SCIP*` pointer. Through some well thought inheritance, we could avoid any code duplication.

> I think just adding a copying step in the reset is a simple fix that would work well at this point.
> For sure it would allow us to ship a clean branching tutorial in time for the Neurips workshop

We can do that, but please move forward in the meantime, because it means I would need to cut another release, which is currently time consuming.

**Comment by dchetelat** [comment]
> On thing that we can perhaps more easily do is have an Instance class that also utilize a SCIP* pointer. Through some well thought inheritance, we could avoid any code duplication.

Sure, I'm not certain how to do that but as long as an instance can be solved several times in a row, I'm happy with it.

> >     I think just adding a copying step in the reset is a simple fix that would work well at this point.
> >     For sure it would allow us to ship a clean branching tutorial in time for the Neurips workshop
> 
> We can do that, but please move forward in the meantime, because it means I would need to cut another release, which is currently time consuming.

Yeah the tutorial has been basically updated, all that's missing is that.

**Comment by gasse** [comment]
Yes I agree with Antoine, some kind of `Instance` class makes sense. Maybe it could simply hold a [`SCIP_PROBDATA`](https://www.scipopt.org/doc/html/type__prob_8h.php#a4802389010c167222c765de0e3e98aef) pointer ?

Maybe there is an elegant solution to our problem, which would avoid a copy altogether. Look at the [SCIPcreateProb](https://www.scipopt.org/doc/html/group__GlobalProblemMethods.php#gaec38a80f593d5159c9a09d5239633f48) function, it takes a method whose purpose is to free the original problem. We could make instances immutable, so that no one ever change an `Instance` object once it is created, and then handle instances with smart pointers as Python does, and we tell SCIP to use a [SCIP_DECL_PROBDELORIG](https://www.scipopt.org/doc/html/probdata__scflp_8c.php#ae183c611c980c39892d5aa107a726b12) that handles this smart pointer. If the problem data is used somewhere else, don't free it.

This is gonna be a bit of work though, so maybe we should stick with a simple problem copy in `reset()` for now.

**Comment by AntoinePrv** [comment]
`env.reset` copy is implemented in 055c1b310b545d90b83a9ae0d909be03d28a59cb

**Comment by gasse** [comment]
Wow, so fast ! Did you implement that in the plane Antoine ? :P
